File: Assignment_3.cpp
#include <iostream>
using namespace std;

class Publication{
public:
    string name;
    float price;

    Publication(){
        name = "";
        price = 0;
    }

    void init(){
        cout<<"Enter the publication's details.."<<endl;
        cout<<"Name of the publication: ";
        cin>>this -> name;
        cout<<"Price of the publication: ";
        cin>>this -> price;
        cout<<endl;
    }

    void displaybook(Publication &p){
        cout<<"Name of the book: "<<p.name<<endl;
        cout<<"Price of the book: "<<p.price<<endl;
    }

    void displaytape(Publication &p){
        cout<<"Name of the tape: "<<p.name<<endl;
        cout<<"Price of the tape: "<<p.price<<endl;
    }
};

class Book:public Publication{
    int pages;

public:
    Book(Publication &pub){
        try{
            cout<<"Enter no of pages present in the book: ";
            int p;
            cin>>p;
            if(p<=0){
                throw p;
            }
            else{
                this -> pages = p;
            }
        }
        catch(int p){
            cout<<"Pages in the book can't be negative or zero"<<endl;
            this -> pages = 0;
            pub.name = "";
            pub.price = 0.0f;
        }
    }

    void display(Publication &p){
        cout<<"Here are the book's details..."<<endl;
        this -> displaybook(p);
        cout<<"No of pages in the book: "<<pages<<endl;
    }
};

class Tape:public Publication{
    float time;

public:
    Tape(Publication &pub){
        try
        {
            float t;
            cout<<"Enter the time length of the tape: ";
            cin>>t;
            if(t<=0){
                throw t;
            }
            else{
                this -> time = t;
            }
        }
        catch(float t)
        {
            cout<<"Time length of the tape can't be negative or null"<<endl;
            pub.name = "";
            pub.price = 0.0f;
            this -> time = 0.0f;
        }
        
    }

    void display(Publication &p){
        cout<<"Here are the tape's details..."<<endl;
        this -> displaytape(p);
        cout<<"Time length of the tape: "<<time<<endl;
    }
};

int main(){
    while(true){
        Publication p;
        p.init();
        cout<<"Enter whether the publication is a book or a tape.."<<endl;
        cout<<"Press b for book and t for tape: ";
        char c;
        cin>>c;        

        if(c=='b'){
            Book b(p);
			cout<<endl;
            b.display(p);
        }

        if(c=='t'){
            Tape t(p);
			cout<<endl;
            t.display(p);
        }

        cout<<"If you want to try again press 1 else press 0: ";
        int choice;
        cin>>choice;
        cout<<endl;

        if(choice==0){
            break;
        }
    }
}

File: Assignment_3.py
"""
RollNo and Name : 21354_Advait Naik

Problem Statement:
Write a python program to compute following computation on matrix:
a) Addition of 2 matrices
b) Subtraction of 2 matrices
c) Multiplication of 2 matrices
d) Transpose of a matrix
"""

#Program
class Matrix:
    def __init__(self,m=None,n=None):
        if(m==None and n==None):
            self.rows = int(input("Enter no of rows in the matrix: "))
            self.columns = int(input("Enter no of columns in the matrix: "))
        else:
            self.rows = m
            self.columns = n
        self.elements = []


    def build(self):
        print("Its a ",self.rows,"X",self.columns," matrix.")
        print("Enter values in the matrix")
        for i in range(self.rows):
            print("Enter values in row",i)
            l1=[]
            for j in range(self.columns):
                value = int(input())
                l1.append(value)
            self.elements.append(l1)

    def display(self):
        print("Entered matrix is as follows")
        for i in range(self.rows):
            for j in range(self.columns):
                print(self.elements[i][j],end = " ")
            print()


    def addition(self,m1):
        Sum = Matrix(self.rows,self.columns)
        if(self.rows==m1.rows and self.columns==m1.columns):
            for i in range(self.rows):
                l1=[]
                for j in range(self.columns):
                    element1 = self.elements[i][j]
                    element2 = m1.elements[i][j]
                    addition = element1 + element2
                    l1.append(addition)
                Sum.elements.append(l1)

            Sum.display()

        else:
            print("Passed matrix should be of dimension ",self.rows,"X",self.columns)



    def subtraction(self,m2):
        Diff = Matrix(self.rows,self.columns)
        if (self.rows == m2.rows and self.columns == m2.columns):
            for i in range(self.rows):
                l2=[]
                for j in range(self.columns):
                    element1 = self.elements[i][j]
                    element2 = m2.elements[i][j]
                    subtraction = element1 - element2
                    l2.append(subtraction)
                Diff.elements.append(l2)

            Diff.display()

        else:
            print("Passed matrix should be of dimension ", self.rows, "X", self.columns)


    def multiplication(self,m3):
        Product = Matrix(self.rows,m3.columns)
        if(self.columns==m3.rows):
            for i in range(self.rows):
                row = []
                for j in range(m3.columns):
                    row.append(0)
                Product.elements.append(row)   

            for i in range(self.rows):
                for j in range(m3.columns):
                    for k in range(m3.rows):
                        Product.elements[i][j] = Product.elements[i][j] + (self.elements[i][k] * m3.elements[k][j])

            Product.display()

        else:
            print("No of columns of 1st matrix doesnt match rows of second matrix.")

    def transpose(self):
        Transpose = Matrix(self.rows,self.columns)
        for i in range(self.rows):
            row = []
            for j in range(self.columns):
                element3 = self.elements[j][i]
                row.append(element3)
            Transpose.elements.append(row)

        Transpose.display()


choice = 1
while(choice==1):
    print("Enter details about 1st Matrix")
    M1 = Matrix()
    print()
    M1.build()
    print()
    print("Enter details about 2nd Matrix")
    M2 = Matrix()
    print()
    M2.build()
    print()
    print("The 2 entered matrix are as follows")
    M1.display()
    print()
    M2.display()
    print()
    n = int(input("Enter choice of operation.... Press 1 for addition,2 for subtraction,3 for multiplictaion "
                  "and 4 for transpose of a matrix: "))
    if(n==1):
        print("The required addition is")
        M1.addition(M2)
    elif(n==2):
        print("The required subtraction is")
        M1.subtraction(M2)
    elif(n==3):
        print("The required multiplication is")
        M1.multiplication(M2)
    elif(n==4):
        option = int(input("Enter the matrix u want to get transpose:"))
        if(option==1):
            M1.transpose()
        elif(option==2):
            M2.transpose()
        else:
            print("Invalid option")
    else:
        print("Invalid choice")

    choice = int(input("Press 1 to do again else press 0 to exit"))